domain-model/src/main/java/uk/gov/hmcts/cmc/domain/constraints/ValidResidence.java

```diff
@@ -0,0 +1,21 @@
+package uk.gov.hmcts.cmc.domain.constraints;
+
+import javax.validation.Constraint;
+import javax.validation.Payload;
+import java.lang.annotation.Documented;
+import java.lang.annotation.ElementType;
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+
+@Target({ElementType.TYPE, ElementType.ANNOTATION_TYPE})
+@Retention(RetentionPolicy.RUNTIME)
+@Constraint(validatedBy = {ValidResidenceConstraintValidator.class})
+@Documented
+public @interface ValidResidence {
```

Comment:
Prefer to create a class level reusable constraint which can validate based on some conditional property. We can apply same on other similar places. 

Something like:
```
@ConditionalRequired(fieldName = "type", values = {"OTHER"}, required = {"otherDetail"})
public class Residence {
    @NotNull
    private final ResidenceType type;
    private final String otherDetail;
...
}
```

and the marker for validator would be like:

```
@Target({ElementType.TYPE})
@Retention(RetentionPolicy.RUNTIME)
@Constraint(validatedBy = {ConditionalRequiredValidator.class})
public @interface ConditionalRequired {

    String message() default "Field is required.";
    Class<?>[] groups() default {};
    Class<? extends Payload>[] payload() default {};

    String fieldName();
    String[] required();
    String[] values();
}
```

Reply:
Then there's no type safety and it'll break at runtime if you refactor, I would prefer against this

Reply:
The unit tests for these models will catch if we miss anything during refactoring. Also we don't change the field names very often. IMHO reusable constraint will be more appropriate and will reduce duplicate code.

Reply:
they may not catch them though, and creating such a validator leads to the chance that it will be used on some optional field which isn't covered fully by tests,
It's unneeded complexity and risk imo

Reply:
Well, some one can still misuse any of the hibernate validators on optional!

I still think `ValidResidence`, `ValidExpense` and `ValidIncome` producing duplication and can be done by one reusable validator. @geek-soft what are your thoughts on this?